### ISSUE TITLE
Non-merged branches listed for deletion?

### ISSUE THREAD

**Issue by felix-antony** [hartwork/git-delete-merged-branches#31]
I was trying to delete remote merged branches, but the script listing non-merged(newly created) branch also for remote delete.  Is there any open issue?  Please help me on this.

**Comment by hartwork** [comment]
Hi @felix-antony , git-delete-merged-branches is not a script :smiley: I'm happy to help you on- or off-GitHub to diagnose the situation. Without a pile of more details, I won't be able to do much.

**Comment by hartwork** [comment]
PS: A newline created branch will show up in the list of branches to delete if and only if the branch points to a commit that has been merged into all the required merge targets that you configured. So likely candidates are (a) no new commits on that branch and (b) unintended mis-configuration. Let's find out :smiley:

**Comment by felix-antony** [comment]
Thanks for your quick reply!
I have create a `test-repo` with default branch `master`, and then `develop` from `master` branch, then created `feature/add-test-feature` branch and `bugfix/fix-emply-lines` branch from `develop` made some commits to those branches, and merged to 'develop` , and then created another custom branch `n-to-merge` branch from develop, but did neither make any commit to this branch nor merge to `develop` .  I cloned this `test-repo` to new location in my computer, and then check-out  `master` and `develop` branches locally, and no other branches checked out locally, and then ran `git-delete-merged-branches` in command-line (screenshot attached for steps which i followed). 
<img width="840" alt="screenshot-1" src="https://user-images.githubusercontent.com/4560562/109808758-84f66b80-7c4d-11eb-9bae-67050ad68652.png">
<img width="574" alt="continue-from-screenshot-1" src="https://user-images.githubusercontent.com/4560562/109808809-92135a80-7c4d-11eb-9c90-f65c522984c7.png">

**Comment by hartwork** [comment]
@felix-antony thanks for sharing these details, I'll play with it and investigate and then get back to you.

**Comment by hartwork** [comment]
Hi @felix-antony, I have followed all your steps in detail by now. The good *and* bad news is that the issue doesn't show here (yet?) — either because (a) you and I are doing slightly different things or (b) because of a difference in version of some dependency of git-delete-merged-branches: Git, Python, etc.

**But** I have turned your description into a Bash script 1:1 and I'll share that script and my output in a dedicated post below. Given that script, it would be awesome if you could *review* if you can spot anything that does not match your description — maybe I missed or misunderstood something.  What do you think?

Best, Sebastian

**Comment by hartwork** [comment]
Script `git-dmb-issue-31-research.sh`:

```bash
#! /usr/bin/env bash
# Copyright (C) 2021 Sebastian Pipping <sebastian@pipping.org>
# Licensed under Creative Commons CC0 1.0 Universal
# https://creativecommons.org/share-your-work/public-domain/cc0/

set -x
PS4="\n${PS4}"  # insert blank lines before xtrace lines
set -eu

cd "$(mktemp -d)"  # not cleaned up, considered a feature here

# "I have create a test-repo with default branch master"
mkdir test-repo
pushd test-repo
git init -b master
git config user.name whoever
git config user.email whoever@whereever.invalid

# "and then develop from master branch"
git commit --allow-empty -m 'Commit on master'
git checkout -b develop

# "then created feature/add-test-feature branch
#  and bugfix/fix-emply-lines branch from develop
#  made some commits to those branches
for branch1 in feature/add-test-feature bugfix/fix-emply-lines ; do
    git checkout -b "${branch1}" develop
    git commit --allow-empty -m "Commit on ${branch1}"
done

# "and merged to develop"
git checkout develop
for branch2 in feature/add-test-feature bugfix/fix-emply-lines ; do
    git merge --no-edit "${branch2}"
done

# "and then created another custom branch n-to-merge branch from develop
#  but did neither make any commit to this branch nor merge to develop"
git branch n-to-merge develop

# "I cloned this test-repo to new location in my computer"
popd
git clone test-repo new-location

# "and then check-out master and develop branches locally
#  and no other branches checked out locally"
cd new-location
git checkout -t origin/master  # develop already exists!

# "and then ran git-delete-merged-branches in command-line"
python3 --version
python3 -m venv venv
source venv/bin/activate
pip install git-delete-merged-branches
git config remote.origin.dmb-enabled true             # We are
for branch3 in master develop ; do                    # faking
    git config "branch.${branch3}.dmb-required" true  # interactive
done                                                  # configuration
git config delete-merged-branches.configured 5.0.0+   # here
cat .git/config
cat /etc/os-release
git --version
git --no-pager branch -a
yes | git-delete-merged-branches  # avoiding -y to fake interactive mode
pwd
```

**Comment by hartwork** [comment]
Output of `git-dmb-issue-31-research.sh` over here:

```console
# ./git-dmb-issue-31-research.sh 
+ PS4='\n+ '

+ set -eu


+ mktemp -d

+ cd /tmp/tmp.wjh97ZQriW

+ mkdir test-repo

+ pushd test-repo
/tmp/tmp.wjh97ZQriW/test-repo /tmp/tmp.wjh97ZQriW

+ git init -b master
Initialized empty Git repository in /tmp/tmp.wjh97ZQriW/test-repo/.git/

+ git config user.name whoever

+ git config user.email whoever@whereever.invalid

+ git commit --allow-empty -m 'Commit on master'
[master (root-commit) 8362d6c] Commit on master

+ git checkout -b develop
Switched to a new branch 'develop'

+ for branch1 in feature/add-test-feature bugfix/fix-emply-lines

+ git checkout -b feature/add-test-feature develop
Switched to a new branch 'feature/add-test-feature'

+ git commit --allow-empty -m 'Commit on feature/add-test-feature'
[feature/add-test-feature f0cca16] Commit on feature/add-test-feature

+ for branch1 in feature/add-test-feature bugfix/fix-emply-lines

+ git checkout -b bugfix/fix-emply-lines develop
Switched to a new branch 'bugfix/fix-emply-lines'

+ git commit --allow-empty -m 'Commit on bugfix/fix-emply-lines'
[bugfix/fix-emply-lines 7974ccf] Commit on bugfix/fix-emply-lines

+ git checkout develop
Switched to branch 'develop'

+ for branch2 in feature/add-test-feature bugfix/fix-emply-lines

+ git merge --no-edit feature/add-test-feature
Updating 8362d6c..f0cca16
Fast-forward

+ for branch2 in feature/add-test-feature bugfix/fix-emply-lines

+ git merge --no-edit bugfix/fix-emply-lines
Already up to date!
Merge made by the 'recursive' strategy.

+ git branch n-to-merge develop

+ popd
/tmp/tmp.wjh97ZQriW

+ git clone test-repo new-location
Cloning into 'new-location'...
done.

+ cd new-location

+ git checkout -t origin/master
Branch 'master' set up to track remote branch 'master' from 'origin'.
Switched to a new branch 'master'

+ python3 --version
Python 3.8.8

+ python3 -m venv venv

+ source venv/bin/activate


+ deactivate nondestructive


+ '[' -n '' ']'


+ '[' -n '' ']'


+ '[' -n /bin/bash -o -n '' ']'


+ hash -r


+ '[' -n '' ']'


+ unset VIRTUAL_ENV


+ '[' '!' nondestructive = nondestructive ']'


+ VIRTUAL_ENV=/tmp/tmp.wjh97ZQriW/new-location/venv


+ export VIRTUAL_ENV


+ _OLD_VIRTUAL_PATH=[..censored..]


+ PATH=[..censored..]


+ export PATH


+ '[' -n '' ']'


+ '[' -z '' ']'


+ _OLD_VIRTUAL_PS1=


+ '[' 'x(venv) ' '!=' x ']'


+ PS1='(venv) '


+ export PS1


+ '[' -n /bin/bash -o -n '' ']'


+ hash -r

+ pip install git-delete-merged-branches
Collecting git-delete-merged-branches
  Using cached git_delete_merged_branches-6.0.5-py3-none-any.whl (35 kB)
Collecting clintermission==0.2.0
  Using cached clintermission-0.2.0-py3-none-any.whl (10 kB)
Collecting colorama>=0.4.3
  Using cached colorama-0.4.4-py2.py3-none-any.whl (16 kB)
Collecting prompt-toolkit
  Using cached prompt_toolkit-3.0.16-py3-none-any.whl (366 kB)
Collecting wcwidth
  Using cached wcwidth-0.2.5-py2.py3-none-any.whl (30 kB)
Installing collected packages: wcwidth, prompt-toolkit, clintermission, colorama, git-delete-merged-branches
Successfully installed clintermission-0.2.0 colorama-0.4.4 git-delete-merged-branches-6.0.5 prompt-toolkit-3.0.16 wcwidth-0.2.5
WARNING: You are using pip version 20.2.3; however, version 21.0.1 is available.
You should consider upgrading via the '/tmp/tmp.wjh97ZQriW/new-location/venv/bin/python3 -m pip install --upgrade pip' command.                                                              

+ git config remote.origin.dmb-enabled true

+ for branch3 in master develop

+ git config branch.master.dmb-required true

+ for branch3 in master develop

+ git config branch.develop.dmb-required true

+ git config delete-merged-branches.configured 5.0.0+

+ cat .git/config
[core]
        repositoryformatversion = 0
        filemode = true
        bare = false
        logallrefupdates = true
[remote "origin"]
        url = /tmp/tmp.wjh97ZQriW/test-repo
        fetch = +refs/heads/*:refs/remotes/origin/*
        dmb-enabled = true
[branch "develop"]
        remote = origin
        merge = refs/heads/develop
        dmb-required = true
[branch "master"]
        remote = origin
        merge = refs/heads/master
        dmb-required = true
[delete-merged-branches]
        configured = 5.0.0+

+ cat /etc/os-release
NAME=Gentoo
ID=gentoo
PRETTY_NAME="Gentoo/Linux"
ANSI_COLOR="1;32"
HOME_URL="https://www.gentoo.org/"
SUPPORT_URL="https://www.gentoo.org/support/"
BUG_REPORT_URL="https://bugs.gentoo.org/"

+ git --version
git version 2.30.1

+ git --no-pager branch -a
  develop
* master
  remotes/origin/HEAD -> origin/develop
  remotes/origin/bugfix/fix-emply-lines
  remotes/origin/develop
  remotes/origin/feature/add-test-feature
  remotes/origin/master
  remotes/origin/n-to-merge

+ yes

+ git-delete-merged-branches
Do you want to run "git remote update --prune" for 1 remote(s):
  - origin

Update? [y/N] 
Do you want to run "git pull --ff-only" for 2 branches(s):
  - develop
  - master

Pull? [y/N] 
No local branches deleted.
No remote branches deleted.

+ pwd
/tmp/tmp.wjh97ZQriW/new-location
```

**Comment by felix-antony** [comment]
Thanks for your reply!
Looking at your output,  I don't see expected result since `bugfix/fix-emply-lines` and `feature/add-test-feature` remote branches are not asked for deletion as those are already merged to develop.

**your output was:**
_No local branches deleted.
No remote branches deleted._

Thanks,
Felix

**Comment by hartwork** [comment]
> Thanks for your reply!
> Looking at your output, I don't see expected result since `bugfix/fix-emply-lines` and `feature/add-test-feature` remote branches are not asked for deletion as those are already merged to develop.
> 
> **your output was:**
> _No local branches deleted. No remote branches deleted._

I know :smiley: You wrote above…

> [..], then created `feature/add-test-feature` branch and `bugfix/fix-emply-lines` branch from `develop` made some commits to those branches, and merged to `develop` , [..]

…so that's what I did. No?

> Hi @felix-antony, I have followed all your steps in detail by now. The good _and_ bad news is that the issue doesn't show here (yet?) — either because (a) you and I are doing slightly different things or (b) because of a difference in version of some dependency of git-delete-merged-branches: Git, Python, etc.

That^^ is key here. Please (a) review my script with regard to a match to your description or (b) turn your scenario into a `Dockerfile` so we can rule out differences in third party dependencies and differences in steps taken. Other ideas? What do you think?

**Comment by hartwork** [comment]
@felix-antony how should we continue? How can I help you make this reproducible?

**Comment by hartwork** [comment]
@felix-antony I have tried on Debian as well just now, previously on Gentoo. There is no reply for 9 days now and without your participitation this is not going anywhere so I'll close as "not reproducible" for now. If you are still interested, I'm happy to re-open and continue research on this, on or off GitHub. My post above outlines [two different ways forward](https://github.com/hartwork/git-delete-merged-branches/issues/31#issuecomment-790579113).
